associations/permissions/marketplace.py

```diff
@@ -1,173 +1,140 @@
-            elif request.method in SAFE_METHODS:
-                return True
-        return False
+            return marketplace.enabled and request.method in SAFE_METHODS
 
 
 class ProductPermission(BasePermission):
     """
-               | Enabled | Disabled |\n
-        Admin  | CRUD    | CRUD     |\n
-        Simple | CRUD    |          |\n\n
-
-        Same as Marketplace.
+               | Enabled | Disabled |
+        Admin  | CRUD    | CRUD     |
+        Simple | R       |          |
     """
 
     message = 'You are not allowed to edit this product.'
-    has_permission = MarketplacePermission.has_permission
+
+    def has_permission(self, request, view):
+        if request.method in ('POST',):
+            market_pk = request.data.get('marketplace', None)
+            market_query = Marketplace.objects.filter(pk=market_pk)
+
+            if market_query.exists():
+                market = market_query[0]
+                role = request.user.get_role(market.association)
+
+                return role and role.marketplace  # Marketplace administrator only.
+            else:
+                raise NotFound('The requested marketplace does not exist.')
+
+        return True
+
+    def has_object_permission(self, request, view, product):
+        role = request.user.get_role(product.marketplace.association)
+
+        if role and role.marketplace:
+            return True  # Marketplace administrator: all the rights.
+        else:
+            return request.method in SAFE_METHODS and product.marketplace.enabled
 
 
 class TransactionPermission(BasePermission):
     """
-               | Enabled | Disabled |\n
-        Admin  | CRU     | CRU      |\n
+               | Enabled | Disabled |
+        Admin  | CRU     | CRU      |
         Simple | CRU     | R        |
     """
 
     message = 'You are not allowed to edit this transaction.'
 
     def has_permission(self, request, view):
-        if request.method in ('DELETE', ):
+        if request.method in SAFE_METHODS:
+            return True
+
+        if request.method in ('DELETE',):
             self.message = 'Transactions cannot be deleted.'
             return False
 
-        role = get_role_in_marketplace(request)
-        if role and role.marketplace:
-            # Marketplace administrator.
+        if request.method in ('POST',):
+            product_pk = request.data.get('product', None)
+            product_query = Product.objects.filter(pk=product_pk)
+
+            if product_query.exists():
+                return product_query[0].marketplace.enabled
+            else:
+                raise NotFound('The requested product does not exist.')
+
+        return True
+
+    def has_object_permission(self, request, view, transaction):
+        # We have to put it a second time here, because when retrieving, both has_permission and has_object_permission
+        # will be called.
+        if request.method in SAFE_METHODS:
```

Comment:
Not everybody should be able to see the transactions of everyone, only themselves and the admins

Reply:
Actually, this is enforced in the view by `get_queryset`.
But maybe it is a bit unclear (and it returns a 404 instead if a 403, which I personnally don't mind).
Do you think I should add a comment? Or explictly forbid in the permission, even if it's redundant?

Reply:
Redundancy is not a bad thing when it comes to such an important security matter

Reply:
Also if for some reason the view came to be changed and did not enforce this, the information would become public

Reply:
OK let's do this